Draggable: Allow elementId based elements to be attached to the ownerDocument body

andrewserong opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Update the `Draggable` component to accept an `appendToOwnerDocument` prop that allows consumers to toggle where `elementId` based elements are attached. It defaults to `false`, and allows setting to `true` to make the element be attached to the `ownerDocument` body instead.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

There are use cases for using an `elementId` based element to clone, but where the consumer needs the cloned element to be attached to the owner document body instead of the element's wrapper (that is, use similar behaviour as exists currently when a custom drag component is used). I encountered this while working on https://github.com/WordPress/gutenberg/pull/49820 where I need the cloned element to be attached to the body instead, so that I can get the draggable chip to render above the drop indicator line.

Why not just use a custom drag component instead? In the case of the list view, we can avoid unnecessarily rendering additional components (e.g. a custom drag chip) by re-using the block list items that are already rendered. Also, the `elementId` logic is very useful, as it already ensures that the cloned element is in the same position as the original element.

However, there are likely lots of cases where styling rules expect the cloned element to be attached to the wrapper. For this reason, I've added the prop, but set it to `false` by default.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

* Add an `appendToOwnerDocument` prop, that defaults to `false`.
* If it is truthy, attach the cloned element to the `ownerDocument.body` instead of the element wrapper (this borrows from the existing behaviour when an `__experimentalDragComponent` is used.
* Add Storybook examples for testing

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

Run storybook locally, and then navigate to the Draggable component:

```
npm run storybook:dev
```

Quick local links to the two added stories:

* Append Element To Owner Document: http://localhost:50240/?path=/story/components-draggable--append-element-to-owner-document (dragging makes the dragged element go over the top of the paragraph)
* Default: http://localhost:50240/?path=/story/components-draggable--default (dragging makes the dragged element go underneath the paragraph)

## Screenshots or screencast <!-- if applicable -->

To demonstrate the difference, I've updated the storybook example so that the wrapper of the draggable element has a `z-index` set that is lower than the `z-index` of the paragraph that flags whether or not the component is being dragged. When `appendToOwnerDocument` is set to `true`, then the cloned draggable element does not receive the z-index of the element wrapper, so the cloned element can go above the paragraph.

| `appendToOwnerDocument` set to `false` | `appendToOwnerDocument` set to `true` |
| --- | --- |
| ![2023-04-20 14 55 31](https://user-images.githubusercontent.com/14988353/233263154-2e9bc7ec-a500-4c5b-96d9-c0fea23009e6.gif) | ![2023-04-20 14 55 54](https://user-images.githubusercontent.com/14988353/233263193-ca12e131-8619-41f5-8229-beec51e8269b.gif) |

andrewserong committed:
Draggable: Allow elementId based elements to be attached to the ownerDocument body

andrewserong committed:
Update changelog

andrewserong committed:
Remove comment

andrewserong committed:
Add missing fullstop in changelog

ramonjd reviewed (approved):
This seems to me like a safe change. I checked storybook only, but it works as advertised and the default reflects behavior already in trunk.

apeatling reviewed (approved):
LGTM :+1:

mirka reviewed (changes requested):
(Temporary block while I review this asap 🙂)

mirka commented on packages/components/src/draggable/README.md:
```diff
@@ -8,9 +8,16 @@ Note that the drag handle needs to declare the `draggable="true"` property and b
 
 The component accepts the following props:
 
+### `attachToElementWrapper`: `boolean`
```
Would it be more intuitive to invert this prop to be `attachToOwnerDocument`? I find it cleaner to avoid true-by-default props whenever possible.

mirka replied:
Also `append` rather than `attach` might be a more standard way to name it.

mirka commented on packages/components/src/draggable/README.md:
```diff
@@ -8,9 +8,16 @@ Note that the drag handle needs to declare the `draggable="true"` property and b
 
+### `attachToElementWrapper`: `boolean`
+
+Whether to attach the cloned element to the element's wrapper when using `elementId`. Setting to false will attach the cloned element to the `ownerDocument` body.
```
I was a bit thrown off by the "when using `elementId`" part because `elementId` is supposed to be a required prop.

```suggestion
Whether to attach the cloned element to the element's wrapper. Setting to false will attach the cloned element to the `ownerDocument` body.
```

andrewserong replied:
I've reworded it, but retained reference to the id, as otherwise it could be confusing that `appendToOwnerDocument` defaults to `false` but is still in use when someone's using the experimental drag component. Happy to keep tweaking / feel free to go with different wording if it still feels off!

mirka commented on packages/components/src/draggable/stories/index.tsx:
```diff
@@ -38,7 +39,7 @@ const DefaultTemplate: ComponentStory< typeof Draggable > = ( args ) => {
 			<div
-				/* eslint-disable no-restricted-syntax, eslint-comments/disable-enable-pair */
+				/* eslint-disable-next-line no-restricted-syntax */
 				id="draggable-example-box"
```
Duplicate ids are not going to work in Docs view (`?path=/docs/components-draggable--default`) where all stories are rendered in the same document.

mirka commented on packages/components/src/draggable/stories/index.tsx:
```diff
@@ -81,3 +82,69 @@ export const Default: ComponentStory< typeof Draggable > = DefaultTemplate.bind(
 );
 Default.args = {};
+
+const AttachElementIdTemplate: ComponentStory< typeof Draggable > = (
```
I don't quite think this is worth the code duplication of adding a separate template. For example, you can add a [story description](https://wordpress.github.io/gutenberg/?path=/docs/components-basecontrol--with-visual-label) in a code comment like this, which eliminates the need for a description rendered inside the story itself:

https://github.com/WordPress/gutenberg/blob/85e2d4e3eaf01073d0011bc2247f5b1c2df516ab/packages/components/src/base-control/stories/index.tsx#L52-L59

mirka commented on packages/components/src/draggable/stories/style.css:
```diff
@@ -0,0 +1,7 @@
+.draggable-storybook-wrapper-element .draggable-storybook-element {
+	background-color: #99faa4;
+}
+
+.draggable-storybook-element {
+	background-color: #fa99c8;
+}
```
I wonder if it'd be better to just have a single `attachToOwnerDocument=true` story that demonstrates the z-index use case (instead of this CSS scoping demo), which is the exact use case that prompted you to implement this prop. Especially because with this CSS stuff, it's kind of impossible to understand what's going on without looking at the source code.

andrewserong replied:
That's a good question — I added the CSS example mostly to demonstrate that the original behaviour was retained (the `true` by default example). However, now that you've suggested going with a new prop that is false by default, there's less of a need for that.

I'll have a play around with `z-index` and see if I can consolidate on a single example that better reflects my use case. Thanks!

andrewserong committed:
Switch to appendToOwnerDocument, simplify stories

andrewserong committed:
Update changelog entry

andrewserong commented:
Thanks for all the detailed feedback @mirka, much appreciated! Lots of good suggestions. I've done the following:

* Updated the param to be called `appendToOwnerDocument` and default to `false`
* Consolidated the storybook examples into a single template, moved the description to a comment
* Updated the `id` to use `useInstanceId` so that the examples work when multiple instances are rendered on the same page (e.g. in the Docs view in Storybook)
* Within the single template, added `z-index` values for the use case that I've been concerned with. This is achieved by:
  * Give the draggable element a wrapper that sets a `z-index` value lower to another element on the page. This allows us to test the default behaviour I ran into when working with the list view.
  * Since we need a test element to work with, I've updated the paragraph to have a `z-index` value that is higher than the draggable element's wrapper. I've also given it a background color and a little padding so that it's clearer what's going on.

The result is that we should now be able to play around with this in Storybook with just the Default story and the Append To Owner Document story. Here's a quick couple of gifs to demonstrate:

| `appendToOwnerDocument` set to `false` | `appendToOwnerDocument` set to `true` |
| --- | --- |
| ![2023-04-20 14 55 31](https://user-images.githubusercontent.com/14988353/233263154-2e9bc7ec-a500-4c5b-96d9-c0fea23009e6.gif) | ![2023-04-20 14 55 54](https://user-images.githubusercontent.com/14988353/233263193-ca12e131-8619-41f5-8229-beec51e8269b.gif) |

Let me know if you'd like to see any other changes!

mirka commented on packages/components/src/draggable/types.ts:
```diff
@@ -17,6 +17,13 @@ export type DraggableProps = {
+	 * Whether to append the cloned element to the `ownerDocument` body.
+	 * By default, elements sourced by id are appended to the element's wrapper.
+	 *
+	 *@default false
```
Formatting nit

```suggestion
	 * @default false
```

mirka reviewed (approved):
Looks great, much cleaner ✨ Thank you for the tweaks!

andrewserong committed:
Fix formatting issue

Co-authored-by: Lena Morita <lena@jaguchi.com>

apeatling commented:
Nice job @mirka on the review!
